lib/foreman/renderer.rb

```diff
@@ -48,20 +51,30 @@ def host_param_false?(name)
 
     def render_safe(template, allowed_methods = [], allowed_vars = {}, scope_variables = {})
       if Setting[:safemode_render]
-        box = Safemode::Box.new self, allowed_methods
-        box.eval(ERB.new(template, nil, '-').src, allowed_vars.merge(scope_variables))
+        box = Safemode::Box.new self, allowed_methods, template_name
+        erb = ERB.new(template, nil, '-')
+        box.eval(erb.src, allowed_vars.merge(scope_variables))
       else
         # we need to keep scope variables and reset them after rendering otherwise they would remain
         # after snippets are rendered in parent template scope
         kept_variables = {}
         scope_variables.each { |k,v| kept_variables[k] = instance_variable_get("@#{k}") }
 
         allowed_vars.merge(scope_variables).each { |k,v| instance_variable_set "@#{k}", v }
-        result = ERB.new(template, nil, '-').result(binding)
+        erb = ERB.new(template, nil, '-')
+        # erb allows to set location since Ruby 2.2
+        erb.location = template_name, 0 if erb.respond_to?(:location=)
+        result = erb.result(binding)
 
         scope_variables.each { |k,v| instance_variable_set "@#{k}", kept_variables[k] }
         result
       end
+
+    rescue ::Racc::ParseError, ::SyntaxError => e
+      # Racc::ParseError is raised in safe mode, SyntaxError in unsafe mode
+      new_e = Foreman::Renderer::SyntaxError.new(N_("Syntax error occurred while parsing the template, make sure you have all ERB tags properly closed and the Ruby syntax is valid. The Ruby error: ") + e.message)
```

Comment:
Use %s in the message and pass `e.message` in as the second arg to the constructor.